### PR TITLE
Roaring optimizations

### PR DESCRIPTION
**Pull request by travisturner**
## Overview
Various fixes that were introduced during the SSB benchmarking project.

### CODE REVIEW

File: roaring/roaring.go [hunk 1]
```diff
@@ -2658,83 +2646,46 @@ func differenceRunRun(a, b *container) *container {
...
+	output := &container{bitmap: make([]uint64, bitmapN), container_type: ContainerBitmap}
 
-		}
+	for i := range a.bitmap {
+		v := a.bitmap[i] & (^b.bitmap[i])
+		output.bitmap[i] = v
+		output.n += int(popcount(v))
+
+	}
+	if output.n < ArrayMaxSize {
```
Comment (jaffee) [inline roaring/roaring.go]: Optimize?

File: roaring/roaring.go [hunk 2]
```diff
@@ -2658,83 +2646,46 @@ func differenceRunRun(a, b *container) *container {
...
-		}
-		if v0 < v1 {
-			output.add(v0)
-			v0, eof0 = itr0.next()
-		} else if v0 > v1 {
-			v1, eof1 = itr1.next()
-		} else {
-			v0, eof0 = itr0.next()
-			v1, eof1 = itr1.next()
+	output := &container{bitmap: make([]uint64, bitmapN), container_type: ContainerBitmap}
```
Comment (jaffee) [inline roaring/roaring.go]: not related to this PR, but we super duper need to rename `container_type`

File: roaring/roaring_internal_test.go [hunk 3]
```diff
@@ -669,24 +668,30 @@ func TestUnionMixed(t *testing.T) {
 	if !reflect.DeepEqual(res.runs, []interval16{{start: 5, last: 10}}) {
 		t.Fatalf("test #3 expected %v, but got %v", []interval16{{start: 5, last: 10}}, res.runs)
 	}
-	c.bitmap = []uint64{0x3}
+
```
Comment (jaffee) [inline roaring/roaring_internal_test.go]: this test should probably be refactored to be table-driven

File: roaring/roaring_internal_test.go [hunk 4]
```diff
@@ -1542,8 +1547,8 @@ func TestDifferenceBitmapRun(t *testing.T) {
 }
 
 func TestDifferenceBitmapArray(t *testing.T) {
-	a := &container{bitmap: make([]uint64, bitmapN)}
-	b := &container{}
+	a := &container{bitmap: make([]uint64, bitmapN), container_type: ContainerBitmap}
```
Comment (jaffee) [inline roaring/roaring_internal_test.go]: This test could use more cases - also the bitmap should use a helper function to construct a full bitmap, and we should make sure we're testing the edges of the container. The helper functions bitmapEvens and bitmapOdds are good for this.

File: roaring/roaring_internal_test.go [hunk 5]
```diff
@@ -2410,3 +2416,83 @@ func TestSearc64(t *testing.T) {
...
+			array:  []uint16{0, 1, 63, 120, 543, 639, 12000, 65534, 65535},
+			bitmap: bitmapEvens(),
+			exp:    []uint16{0, 120, 12000, 65534},
+		},
+	}
+
+	for i, test := range tests {
+		a.array = test.array
+		a.container_type = ContainerArray
+		for i, bmval := range test.bitmap {
```
Comment (jaffee) [inline roaring/roaring_internal_test.go]: we should use helper functions to create full bitmaps and get rid of this gross for loop

File: roaring/roaring_internal_test.go (hunk 4 above)
Comment (tgruben) [reply roaring/roaring_internal_test.go]: there is also a MakeBitmap(..)  that will alloc full bitmap with passed in initialize values

**Pull request merged**